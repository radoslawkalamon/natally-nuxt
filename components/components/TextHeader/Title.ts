import Vue from 'vue'

export default Vue.extend({
  name: 'ComponentsTitle',
  props: {
    title: {
      type: String,
      required: true
    },
    type: {
      type: Number,
      default: 1,
      validator (v) {
        return [1, 2, 3, 4].includes(v)
      }
    },
    shallShowUnderscore: {
      type: Boolean,
      default: true
    }
  },
  computed: {
    titleTag () {
      return `h${this.type}`
    },
    titleClasses () {
      return {
        title: true,
        [`title--type-${this.type}`]: true,
        'title--underscore': this.shallShowUnderscore
      }
    }
  }
})
