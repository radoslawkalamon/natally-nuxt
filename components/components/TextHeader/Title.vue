<template>
  <component
    :is="titleTag"
    :class="titleClasses"
    data-test="components-title"
    v-text="title"
  />
</template>

<script src="./Title.ts" lang="ts"></script>
<style src="./Title.css" scoped></style>
