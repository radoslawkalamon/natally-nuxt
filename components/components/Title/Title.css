.title {
  color: inherit;
  font-family: var(--font-family-sans);
  line-height: var(--line-height-header);
  margin: var(--gap) 0;
}

.title--type-1 {
  font-size: var(--font-size-header-1);
}

.title--type-2 {
  font-size: var(--font-size-header-2);
}

.title--type-3 {
  font-size: var(--font-size-header-3);
}

.title--type-4 {
  font-size: var(--font-size-header-4);
}

.title--underscore::after {
  color: var(--color-accent);
  content: var(--content-after-header);
}
