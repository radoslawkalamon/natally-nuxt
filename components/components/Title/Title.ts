import Vue from 'vue'

const tagValidator = (v: number) => [1, 2, 3, 4].includes(v)

export default Vue.extend({
  name: 'ComponentsTitle',
  props: {
    design: {
      type: Number,
      default: 1,
      validator: tagValidator
    },
    title: {
      type: String,
      required: true
    },
    type: {
      type: Number,
      default: 1,
      validator: tagValidator
    },
    shallShowUnderscore: {
      type: Boolean,
      default: true
    }
  },
  computed: {
    titleTag (): string {
      return `h${this.type}`
    },
    titleClasses (): { [key: string]: boolean } {
      return {
        title: true,
        [`title--design-${this.design}`]: true,
        'title--underscore': this.shallShowUnderscore
      }
    }
  }
})
