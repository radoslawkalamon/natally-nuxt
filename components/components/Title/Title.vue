<template>
  <component
    :is="titleTag"
    :class="titleClasses"
  >
    {{ title }}
  </component>
</template>

<script src="./Title.ts" lang="ts"></script>
<style src="./Title.css" scoped></style>
