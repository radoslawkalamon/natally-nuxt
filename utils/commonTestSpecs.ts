import Vue from 'vue'
import { mount } from '@vue/test-utils'
import { merge } from 'lodash'

export function shallRender (component: typeof Vue, dataTestID: string, defaultOptions?: object): void {
  test('shall render', (): void => {
    const wrapper = mount(component, defaultOptions)
    const isWrapperVisible = wrapper.find(`[data-test="${dataTestID}"]`).isVisible()
    expect(isWrapperVisible).toBeTruthy()
  })
}

export function shallHaveSlot (component: typeof Vue, slot: string, defaultOptions?: object): void {
  test(`shall have slot '${slot}'`, (): void => {
    const text = `This is slot '${slot}' test.`
    const wrapperOptions = merge(defaultOptions, {
      slots: {
        [slot]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHaveStringProp (component: typeof Vue, propName: string, defaultOptions?: object): void {
  test(`shall have string prop '${propName}'`, (): void => {
    const text = `This is prop ${propName} test.`
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHavePropInfluenceOverClassNames (
  component: typeof Vue,
  propName: string,
  propValue: unknown,
  classNameSuffix: string,
  defaultOptions?: object
) {
  test(`shall have ${typeof propValue} prop '${propName}' influence over class names`, () => {
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: propValue
      }
    })
    const wrapper = mount(component, wrapperOptions)
    const isWrapperHaveClass = wrapper.classes().some(cls => cls.includes(classNameSuffix))
    expect(isWrapperHaveClass).toBeTruthy()
  })
}
