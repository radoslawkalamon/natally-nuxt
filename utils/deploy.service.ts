/* eslint-disable no-console */
import { Client, FileInfo, FTPResponse } from 'basic-ftp'
import { ProgressInfo } from 'basic-ftp/dist/ProgressTracker'
import capitalize from './formatters/capitalize'
import spacelize from './formatters/spacelize'

type DeployServiceConfig = {
  host: string;
  port: number;
  username: string;
  password: string;
  remoteCatalog: string;
  remoteIgnoreFile: string[];
  localCatalog: string;
}

type RemoveStrategies = {
  [key in FileInfo['type']]: (name: FileInfo['name']) => Promise<void | FTPResponse>
}

function decoratorLogCall () {
  return function (_target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const targetMethod = descriptor.value

    descriptor.value = function (...args: any[]) {
      const formattedName = spacelize(capitalize(propertyKey))
      console.log(`::: ${formattedName} :::`)
      return targetMethod.apply(this, args)
    }

    return descriptor
  }
}

class DeployService {
  readonly host: string;
  readonly port: number;
  readonly username: string;
  readonly password: string;
  readonly remoteCatalog: string;
  readonly remoteIgnoreFile: string[];
  readonly localCatalog: string;
  FTPClient = new Client()

  constructor (config: DeployServiceConfig) {
    this.host = config.host
    this.port = Number(config.port)
    this.username = config.username
    this.password = config.password
    this.remoteCatalog = config.remoteCatalog
    this.remoteIgnoreFile = config.remoteIgnoreFile
    this.localCatalog = config.localCatalog
  }

  public async start (): Promise<void> {
    try {
      await this.openConnection()
      await this.ensureRemoteCatalog()
      await this.openRemoteCatalog()
      await this.cleanRemoteCatalog()
      await this.uploadToRemoteCatalog()
    } catch (e) {
      console.error(e)
      throw e
    } finally {
      this.closeConnection()
    }
  }

  @decoratorLogCall()
  private async openConnection (): Promise<void> {
    await this.FTPClient.access({
      host: this.host,
      port: this.port,
      user: this.username,
      password: this.password,
      secure: false
    })
  }

  @decoratorLogCall()
  private async ensureRemoteCatalog (): Promise<void> {
    await this.FTPClient.ensureDir(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async openRemoteCatalog (): Promise<void> {
    await this.FTPClient.cd(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async cleanRemoteCatalog (): Promise<void> {
    const remoteFiles = await this.FTPClient.list()

    for (const file of remoteFiles) {
      if (this.remoteIgnoreFile.includes(file.name)) {
        continue
      }

      const removeStrategies: RemoveStrategies = {
        0: async () => await Promise.resolve(),
        1: async name => await this.FTPClient.remove(name),
        2: async name => await this.FTPClient.removeDir(name),
        3: async () => await Promise.resolve()
      }
      await removeStrategies[file.type](file.name)
    }
  }

  @decoratorLogCall()
  private async uploadToRemoteCatalog (): Promise<void> {
    this.FTPClient.trackProgress(this.handlerUploadProgress)
    await this.FTPClient.uploadFromDir(this.localCatalog)
    this.FTPClient.trackProgress()
  }

  private handlerUploadProgress (info: ProgressInfo): void {
    if (info.bytes === 0) {
      process.stdout.write(`Uploading file ${info.name}\n`)
    }
  }

  @decoratorLogCall()
  private closeConnection (): void {
    this.FTPClient.close()
  }
}

export default DeployService
