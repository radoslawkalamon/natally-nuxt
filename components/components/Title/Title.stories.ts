import { Meta, Story } from '@storybook/vue'
import Title from './Title.vue'

const meta: Meta = {
  component: Title,
  title: 'Components / Title',
  argTypes: {
    design: {
      options: [1, 2, 3, 4],
      control: {
        type: 'inline-radio'
      }
    },
    type: {
      options: [1, 2, 3, 4],
      control: {
        type: 'inline-radio'
      }
    }
  }
}
export default meta

const Template: Story = (_args, { argTypes }) => ({
  components: { Title },
  props: Object.keys(argTypes),
  template: '<Title v-bind="$props" />'
})

export const withUnderscore: Story = Template.bind({})
withUnderscore.args = {
  design: 1,
  title: 'Title',
  type: 1,
  shallShowUnderscore: true
}

export const withoutUnderscore: Story = Template.bind({})
withoutUnderscore.args = {
  design: 1,
  title: 'Title',
  type: 1,
  shallShowUnderscore: false
}
