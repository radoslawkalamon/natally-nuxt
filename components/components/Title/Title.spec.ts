import Title from '@/components/components/Title/Title.vue'
import { createDefaultOptionsFactory } from '@/devtools/jest.common.spec.utils'
import { shallPassUnitSanityTest } from '@/devtools/jest.common.spec'

const defaultOptionsFactory = createDefaultOptionsFactory({
  propsData: {
    title: 'Default title'
  }
})

describe('Components / Title', () => {
  describe('Unit', () => {
    shallPassUnitSanityTest({
      component: Title,
      description: 'defaults',
      options: defaultOptionsFactory()
    })

    shallPassUnitSanityTest({
      component: Title,
      description: 'with optional props passed',
      options: defaultOptionsFactory({
        propsData: {
          type: 3,
          design: 4,
          shallShowUnderscore: false
        }
      })
    })

    shallPassUnitSanityTest({
      component: Title,
      description: 'type have no influence over design',
      options: defaultOptionsFactory({
        propsData: {
          type: 4
        }
      })
    })
  })
})
