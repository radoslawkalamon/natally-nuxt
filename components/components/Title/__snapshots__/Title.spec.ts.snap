// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components / Title Unit shall pass unit sanity test | defaults 1`] = `"<h1 class="title title--design-1 title--underscore">Default title</h1>"`;

exports[`Components / Title Unit shall pass unit sanity test | type have no influence over design 1`] = `"<h4 class="title title--design-1 title--underscore">Default title</h4>"`;

exports[`Components / Title Unit shall pass unit sanity test | with optional props passed 1`] = `"<h3 class="title title--design-4">Default title</h3>"`;
